### PR TITLE
Update geopackage integration to support string feature IDs.

### PR DESCRIPTION
**Pull request by jdgarrett**
Since geopackage does not support string feature IDs, the export process will convert all string feature IDs to longs, and then store all of the feature ID mappings in a table in the geopackage.  The import process will check for the existence of this mapping table when bringing in new features and update them with the correct feature IDs.

### CODE REVIEW

**Comment by volaya** [comment]
I tried to use it but cannot import a geopackage. It seems that it searches for a table with the fid mapping, even when importing a geopkg for the first time, so i cannot start the repo from a gpkg file

This is the error from the log

```
2016-06-20 08:39:42,026 ERROR [main] o.l.g.c.GeogigCLI [GeogigCLI.java:389] Unable to insert features into the working tree.
java.sql.SQLException: [SQLITE_ERROR] SQL error or missing database (no such table: countries_fids)
	at org.sqlite.core.DB.newSQLException(DB.java:890) ~[sqlite-jdbc-3.8.6.jar:na]
	at org.sqlite.core.DB.newSQLException(DB.java:901) ~[sqlite-jdbc-3.8.6.jar:na]
	at org.sqlite.core.DB.throwex(DB.java:868) ~[sqlite-jdbc-3.8.6.jar:na]
	at org.sqlite.core.NativeDB.prepare(Native Method) ~[sqlite-jdbc-3.8.6.jar:na]
	at org.sqlite.core.DB.prepare(DB.java:211) ~[sqlite-jdbc-3.8.6.jar:na]
	at org.sqlite.jdbc3.JDBC3Statement.executeQuery(JDBC3Statement.java:81) ~[sqlite-jdbc-3.8.6.jar:na]
	at org.apache.commons.dbcp.DelegatingStatement.executeQuery(DelegatingStatement.java:208) ~[commons-dbcp-1.4.jar:1.4]
	at org.apache.commons.dbcp.DelegatingStatement.executeQuery(DelegatingStatement.java:208) ~[commons-dbcp-1.4.jar:1.4]
	at org.locationtech.geogig.geotools.geopkg.GeopkgGeogigMetadata.getFidMappings(GeopkgGeogigMetadata.java:137) ~[geogig-geotools-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.geotools.geopkg.GeoPkgForwardingFeatureIteratorProvider.forwardIterator(GeoPkgForwardingFeatureIteratorProvider.java:49) ~[geogig-geotools-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.geotools.plumbing.ImportOp$1$1.features(ImportOp.java:402) ~[geogig-geotools-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.geotools.plumbing.ForceTypeAndFidFeatureSource$2.features(ForceTypeAndFidFeatureSource.java:101) ~[geogig-geotools-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.repository.WorkingTree$BlobInsertTask.call(WorkingTree.java:697) ~[geogig-core-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.repository.WorkingTree$BlobInsertTask.call(WorkingTree.java:660) ~[geogig-core-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[na:1.8.0_74]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:1.8.0_74]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:1.8.0_74]
	at java.lang.Thread.run(Unknown Source) ~[na:1.8.0_74]
```

**Comment by volaya** [comment]
Another thing:

I import a shapefile and then export it as a geopackage with the -i option to create the audit tables.

The table with the layer data is exported, but not the audit tables

This is the console output

```
Exporting from countries to countries...
100%
countries exported successfully to countries
Unable to export: Couldn't resolve 'HEAD:countries' to a commitish object
```
And this is the log entry

```
2016-06-20 08:47:10,068 ERROR [main] o.l.g.c.GeogigCLI [GeogigCLI.java:389] Unable to export: Couldn't resolve 'HEAD:countries' to a commitish object
java.lang.IllegalArgumentException: Couldn't resolve 'HEAD:countries' to a commitish object
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:122) ~[guava-18.0.jar:na]
	at org.locationtech.geogig.geotools.geopkg.InterchangeFormat.export(InterchangeFormat.java:165) ~[geogig-geotools-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.geotools.geopkg.GeopkgAuditExport._call(GeopkgAuditExport.java:59) ~[geogig-geotools-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.geotools.geopkg.GeopkgAuditExport._call(GeopkgAuditExport.java:24) ~[geogig-geotools-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.api.AbstractGeoGigOp.call(AbstractGeoGigOp.java:133) ~[geogig-core-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.geotools.cli.geopkg.GeopkgExport.runInternal(GeopkgExport.java:91) ~[geogig-geotools-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.cli.AbstractCommand.run(AbstractCommand.java:68) ~[geogig-cli-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.cli.GeogigCLI.executeInternal(GeogigCLI.java:528) ~[geogig-cli-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.cli.GeogigCLI.execute(GeogigCLI.java:367) ~[geogig-cli-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
	at org.locationtech.geogig.cli.CLI.main(CLI.java:39) [geogig-cli-app-1.0-SNAPSHOT.jar:1.0-SNAPSHOT]
```
The table with the fid mapping is correctly created

This export with -i mode was working fine before, so I guess it's a regression caused by this PR

File: src/geotools/src/main/java/org/locationtech/geogig/geotools/geopkg/GeoPkgForwardingFeatureIteratorProvider.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                if (property instanceof GeometryAttribute) {
+                    builder.set(featureType.getGeometryDescriptor().getName(), property.getValue());
+                } else {
+                    builder.set(property.getName().getLocalPart(), property.getValue());
+                }
+            }
+            Map<Object, Object> userData = next.getUserData();
+            for (Entry<Object, Object> entry : userData.entrySet()) {
+                builder.featureUserData(entry.getKey(), entry.getValue());
+            }
```
Comment (groldan) [inline src/geotools/src/main/java/org/locationtech/geogig/geotools/geopkg/GeoPkgForwardingFeatureIteratorProvider.java]: lines 80 to 90 can be replaced by a simple `builder.init(next)`

File: src/geotools/src/main/java/org/locationtech/geogig/geotools/cli/geopkg/GeopkgExport.java [hunk 2]
```diff
@@ -75,4 +97,39 @@ protected void runInternal(GeogigCLI cli) throws IOException {
...
+                    builder.set(featureType.getGeometryDescriptor().getName(),
+                            property.getValue());
+                } else {
+                    builder.set(property.getName().getLocalPart(), property.getValue());
+                }
+            }
+            Map<Object, Object> userData = feature.getUserData();
+            for (Entry<Object, Object> entry : userData.entrySet()) {
+                builder.featureUserData(entry.getKey(), entry.getValue());
+            }
```
Comment (groldan) [inline src/geotools/src/main/java/org/locationtech/geogig/geotools/cli/geopkg/GeopkgExport.java]: lines 114 to 125 can be replaced by `builder.init((SimpleFeature)feature)`

**Comment by jdgarrett** [comment]
@volaya Did you commit the import of the shapefile?  It should work as long as HEAD resolves to a commit.  The other issues have been addressed by the latest commit.

File: src/geotools/src/main/java/org/locationtech/geogig/geotools/geopkg/GeopkgDataStoreImportOp.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        /**
+         * Import needs to: 1) Import the data 2) Add changes to be staged 3) Commit staged changes
+         */
+        try {
+            // import data into the repository
+            final ImportOp importOp = getImportOp(dataStore);
+            final GeoPackage geopkg = new GeoPackage(geopackage);
+            final DataSource dataSource = geopkg.getDataSource();
+            Connection connection = dataSource.getConnection();
```
Comment (groldan) [inline src/geotools/src/main/java/org/locationtech/geogig/geotools/geopkg/GeopkgDataStoreImportOp.java]: you sure this connection won't leak?

**Pull request merged**